core/primitives/src/stateless_validation/chunk_endorsements_bitmap.rs

```diff
@@ -0,0 +1,211 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use near_primitives_core::types::ShardId;
+use near_schema_checker_lib::ProtocolSchema;
+
+/// Represents a collection of bitmaps, one per shard, to store whether the endorsements from the chunk validators has been received.
+///
+/// For each shard, the endorsements are encoded as a sequence of bits: 1 means endorsement received and 0 means not received.
+/// While the number of chunk validator seats is fixed, the number of chunk-validator assignments may be smaller and may change,
+/// since the seats are assigned to validators weighted by their stake. Thus, we represent the bits as a vector of bytes.
+/// The number of assignments may be less or equal to the number of total bytes. This representation allows increasing
+/// the chunk validator seats in the future (which will be represented by a vector of greater length).
+#[derive(
+    BorshSerialize,
+    BorshDeserialize,
+    serde::Serialize,
+    serde::Deserialize,
+    Debug,
+    Clone,
+    Eq,
+    PartialEq,
+    Default,
+    ProtocolSchema,
+)]
+#[borsh(init=init)]
+pub struct ChunkEndorsementsBitmap {
+    inner: Vec<Vec<u8>>,
+}
+
+/// Type of unsigned integer used in the internal implementation.
+///
+/// The implementation of the bitmap uses an unsigned integer type whose length-in-bits is greater than
+/// the number of chunk validator seats. The number of seats as of the the first implementation was 68,
+/// thus we use u128 for now, but it may change in the future. This does not break the representation
+/// because we do not store the extra bytes when serializing. Instead, we discard the extra bytes for
```

Comment:
Tbh I wouldn't make any assumptions about number of seats (mandates) per shard.
Target mandates only targets "full" mandates. Every node will also have a single "partial" mandate.

Thus with 200 validators, total number of mandates is 68 * 6 (num_shards) + 200 = 608 which is ~100 per shard. Number of chunk_endorsements per shard is currently < 68 only because the same validator always gets many of its mandates assigned to the same shard... So result is the same but the limit is not easy to reason about.

So I would store a length of bitmap as well. Additionally, with increase of number of number of shards, we won't have to revisit this structure.

Reply:
Ok, simplified the implementation and made it independent of any assumption on the size of the assignments. But I do not think we need to record the number of assignments, because when iterating over the bitmap we will know the number of assignments and iterate for that number of iterations (it will return false for the extra bits recorded).